File: codility/cpp/codility_lesson7_2.cpp
// you can use includes, for example:
// #include <algorithm>
#include <stack>

// you can write to stdout for debugging purposes, e.g.
// cout << "this is a debug message" << endl;

int solution(vector<int> &A, vector<int> &B) {
	// write your code in C++14 (g++ 6.2.0)
	stack<int> s;
	int N = A.size();
	int res = 0;

	for (int i = 0; i < N; i++)
	{
		if (B[i] == 0)
		{
			if (s.empty())
				res++;
			else
			{
				while (!s.empty() && s.top() < A[i])
					s.pop();
				if (s.empty())
					res++;
			}
		}
		else if (B[i] == 1)
		{
			s.push(A[i]);
		}
	}

	if (s.empty())
		return res;
	else
		return res + s.size();
}

File: codility/python/codility_lesson7_2.py
# you can write to stdout for debugging purposes, e.g.
# print("this is a debug message")

def solution(A, B):
	# write your code in Python 3.6
	ans = 0
	stack = []
	for a, b in zip(A, B):
		if len(stack) == 0 and b == 0:
			ans += 1
		elif b == 1:
			stack.append(a)
		elif b == 0:
			while len(stack) != 0 and a > stack[-1]:
				stack.pop()
			if len(stack) == 0:
				ans += 1
	ans += len(stack)
	return ans
